aquadoggo/src/replication/strategies/document/diff.rs

```diff
@@ -0,0 +1,307 @@
+        );
+    }
+
+    let mut remote_needs = HashMap::new();
+
+    // Iterate over all local documents and perform a number of state checks in order to
+    // efficiently conclude which operations we should send the remote.
+    //
+    // 1) If the remote doesn't know about our document then we already know they need all
+    //    operations we have.
+    // 2) If they know our document and the view ids are the same then we know the states are
+    //    equal and we don't need to send anything.
+    // 3) If they know our document but we can't calculate their height document height, then they
+    //    are have operations to send us.
+    // 4) If they know our document and we can calculate the height, then compare local and remote
+    //    heights and send operations if they are behind our state.
+    for (document_id, local_document) in local_documents {
+        let remote_document_view_id = remote_documents.get(&document_id);
+
+        if let Some(remote_document_view_id) = remote_document_view_id {
+            // If the local and remote view ids are the same then we know they're at the same
+            // state and we don't need to send anything.
+            if local_document.view_id() == remote_document_view_id {
+                trace!(
+                    "Local and remote document views are equal {} <DocumentViewId {}>: no action required",
+                    document_id.display(),
+                    local_document.view_id().display()
+                );
+                // Continue to loop over the remaining documents.
+                continue;
+            }
+
+            // Calculate the remote document height.
+            let (remote_height, _) =
+                determine_document_height(store, remote_document_view_id).await;
+
+            // If the height for the remote of this document couldn't be calculated then they are
+            // further progressed for this document and so we don't need to send them anything.
+            if remote_height.is_none() {
+                trace!(
+                    "Could not calculate remote document height {} <DocumentViewId {}>: no action required",
+                    document_id.display(),
+                    remote_document_view_id.display()
+                );
+                // Continue to loop over the remaining documents.
+                continue;
+            };
+
+            // Safely unwrap as None case handled above.
+            let remote_height = remote_height.unwrap();
+
+            trace!(
+                "Remote document height: {} <DocumentViewId {}> {remote_height}",
+                local_document.id().display(),
+                remote_document_view_id.display()
+            );
+
+            // Calculate the local document height.
+            //
+            // @TODO: This could be made more efficient if we store the current document height in the
+            // store.
+            let (local_height, _) =
```

Comment:
Ah yeh, missing some logic around here which accounts for document view ids which contain multiple operation ids.

Reply:
When the local node has multiple tips, and the remote doesn't, it can be that the document heights appear to be the same but they're actually not.

Reply:
See comment below: https://github.com/p2panda/aquadoggo/pull/432#issuecomment-1627651760